// @framer/wowr_t.simple-countdown-timer/code/Count.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { ready } from "framer/types/src/render/utils/dom";

const countStyle= {
    fontSize: "50px",
    fontWeight: "bold",
    textAlign: "center"
};

interface Props {
    ms: number;
}

export class Count extends React.Component<Props> {

    static propertyControls: PropertyControls = {
        ms: { type: ControlType.Number, title: "Seconds" },
    }

    render() { 
        // Saving props as variables
        const { ms } = this.props
    }

    constructor(props) {
        super(props);

        this.state = {
            ms: props.ms
        }
    }

    componentDidMount() {
        this.interval = setInterval( () => {
            if ((this.state.ms) <= 0) {
                clearInterval(this.interval);
                this.forceUpdate();
                return;
            }
            this.setState({ms: this.state.ms - 1});
        }, 1000)
    }
    format() {
        const { ms } = this.state;
        let seconds = Math.floor(ms / 1);
        return `${seconds}`;
    }
    render() {
        return (
            <p style={countStyle}>{this.format(this.state.ms)}</p>
        )
    }
}
